sync15-adapter/src/state_machine.rs

```diff
@@ -0,0 +1,599 @@
+                                }
+                            }
+                            for (collection, key_bundle) in &fresh_keys.collections {
+                                if key_bundle != stale_keys.key_for_collection(collection) {
+                                    changes.push(EngineStateChange::Reset(collection.to_string()));
+                                }
+                            }
+                        } else {
+                            // The default bundle changed, so reset all engines
+                            // except those with the same collection-specific
+                            // keys.
+                            let mut except = HashSet::new();
+                            for (collection, key_bundle) in &stale_keys.collections {
+                                if key_bundle == fresh_keys.key_for_collection(collection) {
+                                    except.insert(collection.to_string());
+                                }
+                            }
+                            for (collection, key_bundle) in &fresh_keys.collections {
+                                if key_bundle != stale_keys.key_for_collection(collection) {
+                                    except.insert(collection.to_string());
+                                }
+                            }
+                        }
+                        MetaGlobalEnginesAction::ChangeStates(changes)
+                    },
+                    None => MetaGlobalEnginesAction::ResetAll
+                }
+            },
+            None => MetaGlobalEnginesAction::ResetAll
+        };
+        match action {
+            MetaGlobalEnginesAction::ResetAll => self.add_engine_state_change(EngineStateChange::ResetAll { except: HashSet::new() }),
+            MetaGlobalEnginesAction::ChangeStates(changes) => {
+                for change in changes {
+                    self.add_engine_state_change(change);
+                }
+            }
+        }
+    }
+
+    /// Persists the scratchpad. This method intentionally doesn't return an
+    /// error, since we don't want metadata storage errors to block the client
+    /// from syncing, and we can keep the new `meta/global` and `crypto/keys`
+    /// in memory. (We'll just re-fetch them on the first sync after the
+    /// next startup).
+    fn save(&mut self);
+}
+
+// TODO: Could we generalize this to fetch a token from the token server,
+// and simplify the token server client a bit? If the storage node changes,
+// we'll need to go through the entire process again, but we can skip a few
+// states if the node matches. Maybe we could persist the endpoint and token
+// in the scratchpad (token could be kept in-memory only), in addition to
+// `meta/global` and `crypto/keys`, and add some states for "needs fresh
+// token" and "has fresh token, check if endpoint changed". If thee endpoint is
```

Comment:
s/thee/the/ - but in general, this comment probably makes sense. I guess the same applies for the backoff state?